lib/push.py

```diff
@@ -80,21 +82,63 @@ def manifest_new(class_):
                "layers": [],
                "weirdal": "yankovic" }
 
-   def cleanup(self):
-      ch.INFO("cleaning up")
-      # Delete the tarballs since we can't yet cache them.
-      for (_, tar_c) in self.layers:
-         ch.VERBOSE("deleting tarball: %s" % tar_c)
-         tar_c.unlink_()
+#   def cleanup(self):
+#      ch.INFO("cleaning up")
+#      # Delete the tarballs since we can't yet cache them.
+#      for (_, tar_c) in self.layers:
+#         ch.VERBOSE("deleting tarball: %s" % tar_c)
+#         tar_c.unlink_()
+
+   def pack_prepared_image(self, prepared):
+      """Store a list of gzipped layer tarball paths, config as a sequence of
+       bytes, and manifest as a sequence of bytes."""
+      config = prepared[0]
+      manifest = prepared[1]
+      tars_c = prepared[2]
+      # Pack it up to go.
+      config_bytes = json.dumps(config, indent=2).encode("UTF-8")
+      config_hash = ch.bytes_hash(config_bytes)
+      manifest["config"]["size"] = len(config_bytes)
+      manifest["config"]["digest"] = "sha256:" + config_hash
+      ch.DEBUG("config: %s\n%s" % (config_hash, config_bytes.decode("UTF-8")))
+      manifest_bytes = json.dumps(manifest, indent=2).encode("UTF-8")
+      ch.DEBUG("manifest:\n%s" % manifest_bytes.decode("UTF-8"))
+      # Store for the next steps.
+      self.layers = tars_c
+      self.config = config_bytes
+      self.manifest = manifest_bytes
 
-   def prepare(self):
-      """Prepare self.image for pushing to self.dst_ref. Return tuple: (list
-         of gzipped layer tarball paths, config as a sequence of bytes,
-         manifest as a sequence of bytes).
+   def prepare_existing(self):
+      """Check self.image if an existing config, manifest, and tar exist.
+         Return tuple: (config, manifest, tar) if these files exist."""
+      (sid, git_hash) = bu.cache.find_image(self.image)
+      conf_suffix = git_hash + ".config.json"
+      conf_path = ch.storage.upload_cache // conf_suffix
+      if (conf_path.exists()):
+         config = conf_path.json_from_file("config")
+      else:
+         print("Previous config does not exist")
+         return
+      man_suffix = git_hash + ".manifest.json"
+      man_path = ch.storage.upload_cache // man_suffix
+      if (man_path.exists()):
+         manifest = man_path.json_from_file("manifest")
+      else:
+         print("Previous manifest does not exist")
+         return
+      tar_c = str(self.image.ref) + ".tar.gz"
+      path_c = ch.storage.upload_cache // tar_c
+      if (path_c.exists()):
+         hash_c = path_c.file_hash()
+         tars_c = [(hash_c, path_c)]
+      else:
+         print("Previous tar does not exist")
+         return
+      return (config, manifest, tars_c)
```

Comment:
```suggestion
      self.prepare_upload(config, manifest, tars_c)
```
Instead of returning anything, we just set the class instance attributes.

Reply:
Thanks for the help! This new logic is a lot smarter and simpler.